mlir/lib/Analysis/DataFlow/DenseAnalysis.cpp

```diff
@@ -139,15 +139,43 @@ void AbstractDenseDataFlowAnalysis::visitRegionBranchOperation(
   assert(predecessors->allPredecessorsKnown() &&
          "unexpected unresolved region successors");
 
+  auto getBeforeBranchState = [&](RegionBranchOpInterface branch) {
+    const AbstractDenseLattice *before;
+    if (Operation *prev = branch->getPrevNode())
+      before = getLatticeFor(point, prev);
+    else
+      before = getLatticeFor(point, branch->getBlock());
+    return before;
+  };
+
+  if (auto *op = point.dyn_cast<Operation *>()) {
+    if (op == branch)
+      // In this context, the known predecessors are the last operations along
+      // all execution paths that pass through the branch. If the number of
+      // branch regions is equal to the number of terminates (e.g., scf.yield)
+      // plus one, then we need to join the state before the branch, as the
+      // branch operation doesn't dominate the next operation.
+      // For example (reaching definition):
+      //  store x, p
+      //  scf.if ()
+      //    store y, p
+      //    scf.yield
+      //  load p
+      // => the reaching definition for the load of p should be both stores.
+      // A `scf.if` has 2 regions, the "then" and "else" regions. The "else"
```

Comment:
A `scf.if` operation has always 2 regions (representing the "then" and "else" regions). The "else" region may have 0 blocks, in which case the operation will have only one terminator (the scf.yield on the "then" region).